add: get clients, servers info

leeminju531 opened the pull request:
Add get clients, servers info 

Refer to https://github.com/ros2/ros2cli/pull/916

leeminju531 commented:
**_ref_**
[rmw](https://github.com/ros2/rmw/pull/371)
[rmw_implementation](https://github.com/ros2/rmw_implementation/pull/238)
[rmw_fastrtps](https://github.com/ros2/rmw_fastrtps/pull/771)
[rmw_cyclonedds](https://github.com/ros2/rmw_cyclonedds/pull/499)
[rmw_connextdds](https://github.com/ros2/rmw_connextdds/pull/154)

fujitatomoya commented on rmw/include/rmw/get_topic_endpoint_info.h:
```diff
@@ -186,6 +186,166 @@ rmw_get_subscriptions_info_by_topic(
+ * This function returns an array of endpoint information for each client
+ * of a given service, as discovered so far by the given node.
+ * Endpoint information includes the client's node name and namespace,
+ * the associated topic type, the client's gid, and the client QoS profile.
```
```suggestion
 * the associated service type, the client's gid, and the client QoS profile.
```

fujitatomoya commented on rmw/include/rmw/get_topic_endpoint_info.h:
```diff
@@ -186,6 +186,166 @@ rmw_get_subscriptions_info_by_topic(
+ * of a given service, as discovered so far by the given node.
+ * Endpoint information includes the client's node name and namespace,
+ * the associated topic type, the client's gid, and the client QoS profile.
+ * Names of non-existent topics are allowed, in which case an empty array will be returned.
```
```suggestion
 * Names of non-existent services are allowed, in which case an empty array will be returned.
```

fujitatomoya commented on rmw/include/rmw/get_topic_endpoint_info.h:
```diff
@@ -186,6 +186,166 @@ rmw_get_subscriptions_info_by_topic(
+ * This function returns an array of endpoint information for each server
+ * of a given service, as discovered so far by the given node.
+ * Endpoint information includes the server's node name and namespace,
+ * the associated topic type, the server's gid, and the server QoS profile.
```
```suggestion
 * the associated service type, the server's gid, and the server QoS profile.
```

fujitatomoya commented on rmw/include/rmw/get_topic_endpoint_info.h:
```diff
@@ -186,6 +186,166 @@ rmw_get_subscriptions_info_by_topic(
+ * of a given service, as discovered so far by the given node.
+ * Endpoint information includes the server's node name and namespace,
+ * the associated topic type, the server's gid, and the server QoS profile.
+ * Names of non-existent topics are allowed, in which case an empty array will be returned.
```
```suggestion
 * Names of non-existent services are allowed, in which case an empty array will be returned.
```

fujitatomoya reviewed (changes requested):
I think in rmw service endpoints are handled as topics internally, precisely client is `DataReader` and server is `DataWriter` to fill in the topic endpoint information. so all the field of `rmw_topic_endpoint_info_t` can be filled in with calling the graph cache.

IMO i think in rmw we can use `rmw_topic_endpoint_info_t` as it is now, but i would add comment that says this structure is used for service client and server as well.

sloretz commented:
@wjwwood Mind taking a look at this new rmw API 🧇 ?

methylDragon commented:
@leeminju531 , any updates from your end?

fujitatomoya commented:
@leeminju531 thanks for putting the effort on this.

i will do the another review once they are ready, so please let me know.

leeminju531 commented:
@fujitatomoya @wjwwood 

Sorry for the late response :upside_down_face:


Based on https://github.com/ros2/ros2cli/pull/916#issuecomment-2372588269, we need service information from `rmw`.
I’d like to confirm the structure for representing service information before diving into the implementation details.

I came up with the following structure. Could you take a look and let me know your thoughts?

```C++
typedef struct RMW_PUBLIC_TYPE rmw_service_endpoint_info_s
{
  /// Information about an internally used topic for service communication.
  /// For clients, this field contains a DataReader, while for servers, it holds a DataWriter.
  /// Nullptr for systems that explicitly support services without using topics.
  rmw_topic_endpoint_info_t * internal_topic_info;

  /// The name of the node providing the service.
  const char * node_name;

  /// The namespace of the node providing the service.
  const char * node_namespace;

  /// The type name of the associated service.
  const char * service_type;

  /// A hash representing the service type's description.
  /// Nullptr for systems without explicit service support.
  /// This may be populated in future implementations.
  rosidl_type_hash_t * service_type_hash;

  /// Specifies whether the endpoint is a CLIENT or SERVER.
  rmw_endpoint_type_t endpoint_type;

  /// Globally unique identifier (GID) of the endpoint.
  /// Initialized to zeros (0) for systems without explicit service support.
  /// In future implementations, this may be used to uniquely identify the endpoint.
  uint8_t endpoint_gid[RMW_GID_STORAGE_SIZE];

  /// The QoS profile applied to the service's communication.
  /// This is typically shared for both request and response messages
  /// as they often use the same communication characteristics.
  rmw_qos_profile_t qos_profile;
} rmw_service_endpoint_info_t;
```

fujitatomoya commented:
@leeminju531 glad you came back :smiley: let me have some time, i will bring this up to next PMC meeting.

fujitatomoya commented:
@leeminju531 we had a discussion on the information from `ros2 service info -v` and rmw data structure for this. as in `rmw`, service information structure should not be dependent implementation details, but generic for all implementations. i think we can collect more information from rmw_zenoh what we can do here. and it is expected that service endpoint should have service type hash (not topic type hashes): this one i am not sure if this is implemented as expected... i thought there are only topic type hash is supported, maybe i was wrong on this.

leeminju531 commented:
@fujitatomoya 
Thanks for reviewing this!

Here’s the revised `rmw_service_endpoint_info_t` structure:
```c++
typedef struct RMW_PUBLIC_TYPE rmw_service_endpoint_info_s
{
  /// The name of the node providing the service.
  const char * node_name;

  /// The namespace of the node providing the service.
  const char * node_namespace;

  /// The type name of the associated service.
  const char * service_type;

  /// A hash representing the service type's description.
  rosidl_type_hash_t * service_type_hash;

  /// Specifies whether the endpoint is a CLIENT or SERVER.
  rmw_endpoint_type_t endpoint_type;

  /// The Globally Unique Identifier (GID) of the endpoint.
  /// For systems without explicit service support, this retrieves the Writer's GID.
  uint8_t endpoint_gid[RMW_GID_STORAGE_SIZE];

  /// The Quality of Service (QoS) profile applied to the service's communication.
  /// This is typically shared between the request and response messages as they often use the same communication settings.
  rmw_qos_profile_t qos_profile;
} rmw_service_endpoint_info_t;
```

Modifications: 
- Removed `internal_topic_info` to ensure the structure remains generic across all implementations.
- Updated the comment for `service_type_hash` to clarify that it should be retrieved regardless of the middleware.
- Changed `endpoint_gid` handling: Instead of initializing to zero in DDS-based systems (without explicit service support), it implicitly returns the Writer's GID. This aligns better with how service introspection notifies users of [each client and server GID](https://github.com/ros2/ros2_documentation/blob/jazzy/source/Tutorials/Demos/Service-Introspection.rst?plain=1#L305-L347).

One thing I’m still unsure about is the `qos_profile`. Right now, it assumes that the reader and writer share the same QoS profile, but is a single `qos_profile` sufficient?

PS: I’ll be more available in two months and can dive into the details then.

leeminju531 commented:
Dear @fujitatomoya , @wjwwood 

I implemented a quick prototype based on the definition provided above.
![image](https://github.com/user-attachments/assets/4f56a0f5-ed1f-4d21-abcd-b5faf4497b10)

The definition appears suitable for middleware that explicitly support services, such as Zenoh. 
However, in DDS, several challenges aries : 
1. **Service Type Hash Representation** : Representing the service type hash is problematic. DDS internally addresses services as topics, making it challenging to collect the service type hash within the graph cache. Currently, it seems feasible to retrieve only the topic type hash, not the service type hash.

2. **GID and QoS Profile Representation**: representing the GID and QoS profile is ambiguous. Since DDS uses two GIDs (READER and WRITER) to represent a single service, there is no unique GID for the service itself. Additionally, when using two topics for one service, both aim to write the same QoS profile. However, as far as I know, the middleware will attempt to adapt to the QoS configuration as best it can. Therefore, representing just one QoS profile may not always be accurate.


After further consideration, I revised the structure as follows:
```c++
typedef struct RMW_PUBLIC_TYPE rmw_service_endpoint_info_s
{
/// Name of the node
const char * node_name;
/// Namespace of the node
const char * node_namespace;
/// The associated service type's name
const char * service_type;
/// The endpoint type. (CLIENT or SERVER)
rmw_endpoint_type_t endpoint_type;
/// Specifies whether the middleware explicitly supports services (e.g., Zenoh).
/// The `endpoint_count` value is determined as follows:
/// - 1 if the middleware explicitly supports services.
/// - 2 if request/response are represented as separate reader/writer topics.
size_t endpoint_count;
/// Hashed value for service type's description
rosidl_type_hash_t * service_type_hashes;
/// The GID of the endpoint
uint8_t (*endpoint_gids)[RMW_GID_STORAGE_SIZE];
/// QoS profile of the endpoint
rmw_qos_profile_t * qos_profiles;
} rmw_service_endpoint_info_t;
```
In this revised structure, I introduced the `endpoint_count` field. For middleware that explicitly supports services, it would be set to 1, while for those that do not (e.g., using separate topics for requests and responses like DDS), it would be 2.

Here’s the output when using Zenoh (an explicit service middleware):
![image](https://github.com/user-attachments/assets/8c754390-7307-45f5-bf4f-11eeb31d490b)

If the middleware does not support explicit services, the output would look like this:
```c++
Node name: minimal_service
Node namespace: /
Service type: example_interfaces/srv/AddTwoInts
Service type hashes: 
 - RIHS01_e118de6bf5eeb66a2491b5bda11202e7b68f198d6f67922cf30364858239c81a
 - RIHS01_e118de6bf5eeb66a2491b5bda11202e7b68f198d6f67922cf30364858239c81b
Endpoint type: SERVER
Endpoint count: 2
GIDs: 
 - c1.e9.6a.12.53.b3.5e.28.c7.25.0b.30.36.53.8a.ad
 - c1.e9.6a.12.53.b3.5e.28.c7.25.0b.30.36.53.8a.ae
QoS profiles:
 - Reliability: RELIABLE
   History (Depth): KEEP_LAST (10)
   Durability: VOLATILE
   Lifespan: Infinite
   Deadline: Infinite
   Liveliness: AUTOMATIC
   Liveliness lease duration: Infinite
 - Reliability: RELIABLE
   History (Depth): KEEP_LAST (10)
   Durability: VOLATILE
   Lifespan: Infinite
   Deadline: Infinite
   Liveliness: AUTOMATIC
   Liveliness lease duration: Infinite

```
Please let me know if you need further adjustments or improvements. I would greatly appreciate any feedback or suggestions to ensure it is as effective as possible. Thanks!

fujitatomoya commented:
@leeminju531 thanks for the update and checking in details. images are broken, especially for zenoh one, could you attach it again?

fujitatomoya commented:
> DDS internally addresses services as topics, making it challenging to collect the service type hash within the graph cache. Currently, it seems feasible to retrieve only the topic type hash, not the service type hash.

i think service type needs to be collected from service typesupport, and that is what rmw_zenoh does with `graph_services_`.
probably DDS-based rmw implementation also needs to have graph extension for services, service type hash needs to be consistent between all rmw implementations?
because topic type hash is not the service type hash generated by rosidl, and that is the wrong information for user.

do you happen to have any thought how to manage service type hash in the graph for DDS rmws?

> GID and QoS Profile Representation: representing the GID and QoS profile is ambiguous. Since DDS uses two GIDs (READER and WRITER) to represent a single service, there is no unique GID for the service itself. Additionally, when using two topics for one service, both aim to write the same QoS profile. However, as far as I know, the middleware will attempt to adapt to the QoS configuration as best it can. Therefore, representing just one QoS profile may not always be accurate.

this is true. Requested QoS and actual one can be different, and there is no unique service entity GID...
having `endpoint_count` in `rmw_service_endpoint_info_t` looks good to me, but maybe it should print which one is reader or writer as output?

leeminju531 commented:
> this is true. Requested QoS and actual one can be different, and there is no unique service entity GID...
> having endpoint_count in rmw_service_endpoint_info_t looks good to me, but maybe it should print which one is reader or writer as output?

To handle READER/WRITER distinction, I propose we implicitly guide the interpretation by defining the order clearly in the comments.
```c++
typedef struct RMW_PUBLIC_TYPE rmw_service_endpoint_info_s
{
...
/// GIDs of the endpoints.
/// If `endpoint_count` is 2, ordered as [REQUEST_READER, RESPONSE_WRITER].
uint8_t (*endpoint_gids)[RMW_GID_STORAGE_SIZE];
/// QoS profile of the endpoint
/// If `endpoint_count` is 2, ordered as [REQUEST_READER, RESPONSE_WRITER].
rmw_qos_profile_t * qos_profiles;
...
} rmw_service_endpoint_info_t;
```
Then in rclpy or rclcpp, we can explicitly print or represent which is READER or WRITER by reading the order.
This keeps the RMW layer generic while still providing the necessary info for higher-level tools. 
<br />


> i think service type needs to be collected from service typesupport, and that is what rmw_zenoh does with graph_services_.
> probably DDS-based rmw implementation also needs to have graph extension for services, service type hash needs to be consistent between all rmw implementations?
> because topic type hash is not the service type hash generated by rosidl, and that is the wrong information for user.
> 
> do you happen to have any thought how to manage service type hash in the graph for DDS rmws?

Yeah, I agree. I’ll look into it, but it’d be super helpful if someone could share some guidance on how this should be handled properly.

fujitatomoya commented:
hmmm, yeah this is hard to come up with good solution. i mean we can add service type hash and graph to the DDS-based rmw implementation like rmw_zenoh, but it always comes back to QoS problem...

fujitatomoya commented:
@leeminju531 so we talked about this at ROS PMC once again.

- (we already know) single service type hash needs to be printed with `ros2 service info -v` on service. that means we need to add service type hash identification from ROS 2 to graph cache in DDS based rmw implementations. (I am not sure what the best way to do this, but let's discuss on that together.)
- showing 2 QoS (maybe print one QoS if they are identical) and GIDs with `endpoint_count` is fine, but probably we do not need to print internal topic type hash identifications since those are hidden behind RMW implementation details.
- we can show 2 QoS and GID with DDS based RMWs, but actually there is no way for user to specify 2 QoS configurations for the service. we only can set QoS for either `service` client or server, we cannot set the different QoS to request topic and response topic separately. but this does not mean we should not print what we can show as information.

what do you think?

ros-discourse commented:
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-pmc-minutes-for-april-22-2025/43346/1

leeminju531 commented:
@fujitatomoya , Thanks for bringing this up again. 

As far as I understand, the previous definition is generally acceptable, except for internal topic type hashes. 
I now understand that showing the `topic type hash` is part of the implementation details—thanks for clarifying that.

Additionally, we could explore a cleaner or more user-friendly way to display this information in ros2cli at the rclpy layer like [this](https://github.com/ros2/rclpy/blob/3414456ddaf1163d1951d301336028c9600bd58e/rclpy/rclpy/topic_endpoint_info.py#L187-L208).

Right now, the challenge I’m facing is retrieving the `service_type_hash` from the graph cache. It's more complicated than I expected. I’m still looking into it and digging deeper. Once I figure it out, I’ll follow up!

leeminju531 committed:
Add clients and servers info by service

Signed-off-by: Minju, Lee <dlalswn531@naver.com>

ahcorde commented on rmw/include/rmw/service_endpoint_info.h:
```diff
@@ -0,0 +1,426 @@
+  /// The `endpoint_count` value is determined as follows:
+  /// - 1 if the middleware explicitly supports services (e.g., Zenoh).
+  /// - 2 if request/response are represented as separate reader/writer topics (e.g., DDS).
+  size_t endpoint_count;
```
include `<stddef.h>`

ahcorde commented on rmw/include/rmw/service_endpoint_info.h:
```diff
@@ -0,0 +1,426 @@
+  /// The GIDs of the endpoint(s). The array size is `endpoint_count`.
+  /// If the middleware represents the request and response as separate topics,
+  /// the GIDs should be provided sequentially in the order: reader, writer.
+  uint8_t (*endpoint_gids)[RMW_GID_STORAGE_SIZE];
```
#include `<stdint.h>`

ahcorde commented on rmw/include/rmw/service_endpoint_info_array.h:
```diff
@@ -0,0 +1,155 @@
+typedef struct RMW_PUBLIC_TYPE rmw_service_endpoint_info_array_s
+{
+  /// Size of the array.
+  size_t size;
```
include `<stddef.h>`

ahcorde commented on rmw/src/service_endpoint_info.c:
```diff
@@ -0,0 +1,325 @@
+rmw_ret_t
+rmw_service_endpoint_info_set_gids(
+  rmw_service_endpoint_info_t * service_endpoint_info,
+  const uint8_t * gids,
+  size_t endpoint_count,
```
include `<stdint.h>` and `<stddef.h>`

ahcorde commented on rmw/src/service_endpoint_info.c:
```diff
@@ -0,0 +1,325 @@
+    RMW_SET_ERROR_MSG("failed to allocate memory for endpoint_gids");
+    return RMW_RET_BAD_ALLOC;
+  }
+  memset(service_endpoint_info->endpoint_gids, 0, RMW_GID_STORAGE_SIZE * endpoint_count);
```
include `<string.h>`

ahcorde commented on rmw/src/service_endpoint_info_array.c:
```diff
@@ -0,0 +1,97 @@
+    return RMW_RET_BAD_ALLOC;
+  }
+  service_endpoint_info_array->size = size;
+  for (size_t i = 0; i < size; i++) {
```
include `<stddef.h>`

ahcorde commented on rmw/test/test_service_endpoint_info.cpp:
```diff
@@ -0,0 +1,458 @@
+
+char * get_mallocd_string(const char * s)
+{
+  size_t size = strlen(s) + 1;
```
`<stddef.h>` and `<string.h>`

ahcorde commented on rmw/test/test_service_endpoint_info.cpp:
```diff
@@ -0,0 +1,458 @@
+    rmw_get_zero_initialized_service_endpoint_info();
+  rosidl_type_hash_t type_hash = rosidl_get_zero_initialized_type_hash();
+  type_hash.version = 22;
+  for (uint8_t i = 0; i < ROSIDL_TYPE_HASH_SIZE; i++) {
```
include `<stdint.h>`

leeminju531 committed:
Fix headers from review

Signed-off-by: Minju, Lee <dlalswn531@naver.com>

ahcorde commented:
Pulls: ros2/rmw#371, ros2/rmw_implementation#238, ros2/rcl#1161, ros2/rclcpp#2569, ros2/rclpy#1307, ros2/rmw_zenoh#679, ros2/rmw_fastrtps#771, ros2/rmw_cyclonedds#499, ros2/rmw_connextdds#154, ros2/ros2cli#916
Gist: https://gist.githubusercontent.com/ahcorde/302c3652a6220b1c87e62745e9f9afff/raw/742f65fffc9bfec58f851df6062324b9d1a85a3c/ros2.repos
BUILD args:  --packages-above-and-dependencies rmw rmw_implementation rcl rclcpp rclpy ros2cli ros2service
TEST args:  --packages-above rmw rmw_implementation rcl rclcpp rclpy ros2cli ros2service
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/16548
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=24390)](http://ci.ros2.org/job/ci_linux/24390/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=18457)](http://ci.ros2.org/job/ci_linux-aarch64/18457/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=3835)](http://ci.ros2.org/job/ci_linux-rhel/3835/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=24953)](http://ci.ros2.org/job/ci_windows/24953/)

fujitatomoya commented:
@Mergifyio rebase

fujitatomoya commented on rmw/include/rmw/get_service_endpoint_info.h:
```diff
@@ -0,0 +1,228 @@
+// Copyright 2025 Minju Lee (이민주). All rights reserved.
```
we can attribute ourselves, but "all right reserved" is extra. see appendix on https://opensource.org/license/apache-2-0

```suggestion
// Copyright 2025 Minju Lee (이민주).
```

fujitatomoya commented on rmw/include/rmw/service_endpoint_info.h:
```diff
@@ -0,0 +1,426 @@
+// Copyright 2025 Minju Lee (이민주). All rights reserved.
```
we can attribute ourselves, but "all right reserved" is extra. see appendix on https://opensource.org/license/apache-2-0

```suggestion
// Copyright 2025 Minju Lee (이민주).
```

fujitatomoya commented on rmw/include/rmw/service_endpoint_info_array.h:
```diff
@@ -0,0 +1,157 @@
+// Copyright 2025 Minju Lee (이민주). All rights reserved.
```
we can attribute ourselves, but "all right reserved" is extra. see appendix on https://opensource.org/license/apache-2-0

```suggestion
// Copyright 2025 Minju Lee (이민주).
```

fujitatomoya commented on rmw/src/service_endpoint_info.c:
```diff
@@ -0,0 +1,329 @@
+// Copyright 2025 Minju Lee (이민주). All rights reserved.
```
we can attribute ourselves, but "all right reserved" is extra. see appendix on https://opensource.org/license/apache-2-0

```suggestion
// Copyright 2025 Minju Lee (이민주).
```

fujitatomoya commented on rmw/test/test_service_endpoint_info.cpp:
```diff
@@ -0,0 +1,458 @@
+// Copyright 2025 Minju Lee (이민주). All rights reserved.
```
we can attribute ourselves, but "all right reserved" is extra. see appendix on https://opensource.org/license/apache-2-0

```suggestion
// Copyright 2025 Minju Lee (이민주).
```

fujitatomoya commented on rmw/test/test_service_endpoint_info_array.cpp:
```diff
@@ -0,0 +1,106 @@
+// Copyright 2025 Minju Lee (이민주). All rights reserved.
```
we can attribute ourselves, but "all right reserved" is extra. see appendix on https://opensource.org/license/apache-2-0

```suggestion
// Copyright 2025 Minju Lee (이민주).
```

fujitatomoya commented:
Pulls: ros2/ros2cli#916, ros2/rmw#371, ros2/rmw_implementation#238, ros2/rmw_fastrtps#771, ros2/rmw_cyclonedds#499, ros2/rmw_connextdds#154, ros2/rcl#1161, ros2/rclcpp#2569, ros2/rclpy#1307, ros2/rmw_dds_common#82, ros2/rmw_zenoh#679
Gist: https://gist.githubusercontent.com/fujitatomoya/01385d70469b48f205a68e21219fd395/raw/b02c84fd17938124e2b045dbf016f260e6209ae0/ros2.repos
BUILD args:  --packages-above-and-dependencies rmw_zenoh_cpp rmw_dds_common rclpy rclcpp_lifecycle rclcpp rcl rmw_connextdds rmw_connextdds_common rmw_connextddsmicro rmw_cyclonedds_cpp rmw_fastrtps_cpp rmw_fastrtps_dynamic_cpp rmw_fastrtps_sharedcpp ros2cli ros2service rmw rmw_implementation test_rmw_implementation
TEST args:  --packages-above rmw_zenoh_cpp rmw_dds_common rclpy rclcpp_lifecycle rclcpp rcl rmw_connextdds rmw_connextdds_common rmw_connextddsmicro rmw_cyclonedds_cpp rmw_fastrtps_cpp rmw_fastrtps_dynamic_cpp rmw_fastrtps_sharedcpp ros2cli ros2service rmw rmw_implementation test_rmw_implementation
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/17407
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=25617)](http://ci.ros2.org/job/ci_linux/25617/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=19649)](http://ci.ros2.org/job/ci_linux-aarch64/19649/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=5175)](http://ci.ros2.org/job/ci_linux-rhel/5175/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25848)](http://ci.ros2.org/job/ci_windows/25848/)

fujitatomoya commented:
Pulls: ros2/ros2cli#916, ros2/rmw#371, ros2/rmw_implementation#238, ros2/rmw_fastrtps#771, ros2/rmw_cyclonedds#499, ros2/rmw_connextdds#154, ros2/rcl#1161, ros2/rclcpp#2569, ros2/rclpy#1307, ros2/rmw_dds_common#82, ros2/rmw_zenoh#679
Gist: https://gist.githubusercontent.com/fujitatomoya/3c50929d3b3e884ba6796f6c9f26e92e/raw/b02c84fd17938124e2b045dbf016f260e6209ae0/ros2.repos
BUILD args:  --packages-above-and-dependencies rmw_zenoh_cpp rmw_dds_common rclpy rclcpp_lifecycle rclcpp rcl rmw_connextdds rmw_connextdds_common rmw_connextddsmicro rmw_cyclonedds_cpp rmw_fastrtps_cpp rmw_fastrtps_dynamic_cpp rmw_fastrtps_shared_cpp ros2cli ros2service rmw rmw_implementation test_rmw_implementation
TEST args:  --packages-above rmw_zenoh_cpp rmw_dds_common rclpy rclcpp_lifecycle rclcpp rcl rmw_connextdds rmw_connextdds_common rmw_connextddsmicro rmw_cyclonedds_cpp rmw_fastrtps_cpp rmw_fastrtps_dynamic_cpp rmw_fastrtps_shared_cpp ros2cli ros2service rmw rmw_implementation test_rmw_implementation
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/17408
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=25618)](http://ci.ros2.org/job/ci_linux/25618/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=19650)](http://ci.ros2.org/job/ci_linux-aarch64/19650/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=5176)](http://ci.ros2.org/job/ci_linux-rhel/5176/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25849)](http://ci.ros2.org/job/ci_windows/25849/)

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=25628)](http://ci.ros2.org/job/ci_linux/25628/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=19661)](http://ci.ros2.org/job/ci_linux-aarch64/19661/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=5184)](http://ci.ros2.org/job/ci_linux-rhel/5184/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25857)](http://ci.ros2.org/job/ci_windows/25857/)

leeminju531 committed:
Delete restricting right comments

Signed-off-by: Minju, Lee <dlalswn531@naver.com>

Yadunund commented:
@ahcorde @fujitatomoya the CI result above shows test failures for `test_rmw_implementation.TestGraphAPI.get_servers_info_by_service_with_bad_arguments`. Are any further changes needed?

leeminju531 commented:
@Yadunund, Those failures were fixed at https://github.com/ros2/ros2cli/pull/916#issuecomment-3509406823.

fujitatomoya commented:
@Yadunund besides that, we are verifying this with full CI on https://github.com/ros2/ros2cli/pull/916#issuecomment-3510999380

ahcorde commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=25650)](http://ci.ros2.org/job/ci_linux/25650/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=19702)](http://ci.ros2.org/job/ci_linux-aarch64/19702/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=5497)](http://ci.ros2.org/job/ci_linux-rhel/5497/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=25879)](http://ci.ros2.org/job/ci_windows/25879/)
